File: Java/Hyperverse Server/src/main/java/uk/ac/ucl/hvtp/Packet.java
package uk.ac.ucl.hvtp;

import java.io.Serializable;
import java.util.Objects;

public class Packet implements Serializable {
	private final PacketHeader header;
	private final byte[] payload;

	public Packet(PacketHeader header, byte[] payload) {
		Objects.requireNonNull(header);
		Objects.requireNonNull(payload);

		this.header = header;
		this.payload = payload;
	}

	public PacketHeader getHeader() {
		return header;
	}

	public byte[] getPayload() {
		return payload;
	}

	public String toString() {
		return header.toString();
	}

	public static Packet newInitPacket(byte[] payload) {
		return new Packet(new PacketHeader(HVTPConstants.MAGIC, HVTPConstants.VERSION_1, payload.length, HVTPConstants.INIT), payload);
	}

}

File: Unity/Hyperverse Client/Assets/Packet.cs
using System;
using HVTP; 

namespace HVTP
{
    public class Packet
    {
        private PacketHeader header;
        private byte[] payload;

        public Packet(PacketHeader header, byte[] payload)
        {
            this.header = header;
            this.payload = payload;
        }

        public PacketHeader GetHeader()
        {
            return header;
        }

        public byte[] GetPayload()
        {
            return payload;
        }
    }

    public class PacketHeader
    {
    	public static readonly int SIZE_IN_BYTES = 16;
 
        private string magic;
        private int version;
        private string type;
        private int length;

        public PacketHeader(string magic, int version, string type, int length)
        {
            this.magic = magic;
            this.version = version;
            this.type = type;
            this.length = length;
        }

        public string GetMagic()
        {
            return magic;
        }

        public int GetVersion()
        {
            return version;
        }

        public string GetPacketType()
        {
            return type;
        }

        public int GetLength()
        {
            return length;
        }
    }
}